File: backend/src/services/riskAlertService.js
// src/services/riskAlertService.js
// O ÚNICO caminho pelo qual o Suporte chega a uma conversa do Chat Comercial.
//
// Nada aqui dá ao Suporte acesso a "todas as conversas comerciais" — ver a
// regra crítica do pedido: só as sinalizadas (RiskAlert ativo) aparecem, e
// aceder a UMA delas fica sempre gravado na auditoria. requirePermission
// (SUPORTE) já é aplicado nas rotas antes de chegar aqui; este serviço
// verifica a segunda condição, que a rota não sabe verificar sozinha: que a
// conversa TEM MESMO um alerta.
const prisma = require('../config/database');
const { NotFoundError, ValidationError } = require('../utils/errors');
const auditService = require('./auditService');

const STATUS_RECLASSIFICACAO = ['EM_ANALISE', 'FALSO_POSITIVO', 'RESOLVIDO'];
const ACTION_POR_STATUS = {
  EM_ANALISE: 'ALERTA_SEGURANCA_EM_ANALISE',
  FALSO_POSITIVO: 'ALERTA_SEGURANCA_FALSO_POSITIVO',
  RESOLVIDO: 'ALERTA_SEGURANCA_RESOLVIDO',
};

async function listarAlertas({ status } = {}) {
  const where = status ? { status } : { status: { in: ['ABERTO', 'EM_ANALISE'] } };
  const alertas = await prisma.riskAlert.findMany({ where, orderBy: { createdAt: 'desc' }, take: 100 });
  if (!alertas.length) return [];
  const companyIds = new Set();
  const conversas = await prisma.conversation.findMany({ where: { id: { in: alertas.map((a) => a.conversationId) } } });
  const convPorId = Object.fromEntries(conversas.map((c) => [c.id, c]));
  conversas.forEach((c) => { companyIds.add(c.buyerCompanyId); companyIds.add(c.supplierCompanyId); });
  const empresas = await prisma.company.findMany({ where: { id: { in: [...companyIds] } }, select: { id: true, name: true } });
  const empPorId = Object.fromEntries(empresas.map((e) => [e.id, e.name]));
  return alertas.map((a) => {
    const c = convPorId[a.conversationId];
    return {
      ...a,
      conversation: c ? {
        id: c.id, buyerCompanyId: c.buyerCompanyId, supplierCompanyId: c.supplierCompanyId,
        buyerCompany: empPorId[c.buyerCompanyId], supplierCompany: empPorId[c.supplierCompanyId],
      } : null,
    };
  });
}

// Acesso a UMA conversa sinalizada — exige um alerta (qualquer estado, mesmo
// já resolvido: o histórico de uma conversa que já teve alerta continua a ser
// legítimo consultar). Sem alerta nenhuma vez, nem o Suporte entra: é a
// fronteira que torna a exceção "só o que tem risco relevante", real.
async function acederConversaSinalizada(conversationId, admin, req) {
  const temAlerta = await prisma.riskAlert.findFirst({ where: { conversationId } });
  if (!temAlerta) throw new NotFoundError('Conversa sinalizada');
  const conversa = await prisma.conversation.findUnique({ where: { id: conversationId } });
  if (!conversa) throw new NotFoundError('Conversa sinalizada');
  const [mensagens, alertas] = await Promise.all([
    prisma.conversationMessage.findMany({ where: { conversationId }, orderBy: { createdAt: 'asc' } }),
    prisma.riskAlert.findMany({ where: { conversationId }, orderBy: { createdAt: 'desc' } }),
  ]);
  await auditService.recordSafe({
    actor: auditService.actorFrom(req), action: 'CONVERSA_SINALIZADA_ACEDIDA',
    entityType: 'Conversation', entityId: conversationId,
    detail: auditService.contextoFrom(req),
  });
  return { conversation: conversa, messages: mensagens, alerts: alertas };
}

async function reclassificar(alertId, admin, { status, decision }, req) {
  if (!STATUS_RECLASSIFICACAO.includes(status)) throw new ValidationError('Estado de alerta inválido.');
  const alerta = await prisma.riskAlert.findUnique({ where: { id: alertId } });
  if (!alerta) throw new NotFoundError('Alerta de segurança');
  const atualizado = await prisma.riskAlert.update({
    where: { id: alertId },
    data: { status, decision: String(decision || '').slice(0, 1000) || null, reviewedById: admin.id, reviewedAt: new Date() },
  });
  await auditService.recordSafe({
    actor: auditService.actorFrom(req), action: ACTION_POR_STATUS[status],
    entityType: 'RiskAlert', entityId: alertId,
    detail: { ...auditService.contextoFrom(req), conversationId: alerta.conversationId, decision: decision || null },
  });
  return atualizado;
}

module.exports = { listarAlertas, acederConversaSinalizada, reclassificar };

File: backend-java/src/main/java/ao/kixima/risk/RiskAlertService.java
package ao.kixima.risk;

import ao.kixima.audit.AuditService;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.error.ValidationException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.conversation.Conversation;
import ao.kixima.conversation.ConversationMessage;
import ao.kixima.conversation.ConversationMessageRepository;
import ao.kixima.conversation.ConversationRepository;
import ao.kixima.security.CurrentUser;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Espelha backend/src/services/riskAlertService.js — o ÚNICO caminho pelo
 * qual o Suporte chega a uma conversa do Chat Comercial. Nada aqui dá ao
 * Suporte acesso a "todas as conversas comerciais": só as sinalizadas
 * (RiskAlert ativo) aparecem, e aceder a UMA delas fica sempre gravado na
 * auditoria. RBAC (ADMIN_SISTEMA + área Suporte) é aplicado ao nível do
 * controller ({@code @RequireRole}/{@code @RequirePermission}) — este
 * serviço verifica só a segunda condição, que a rota não sabe verificar
 * sozinha: que a conversa TEM MESMO um alerta.
 */
@Service
public class RiskAlertService {

    private static final Set<RiskAlertStatus> STATUS_RECLASSIFICACAO =
            Set.of(RiskAlertStatus.EM_ANALISE, RiskAlertStatus.FALSO_POSITIVO, RiskAlertStatus.RESOLVIDO);

    private static final Map<RiskAlertStatus, String> ACTION_POR_STATUS = Map.of(
            RiskAlertStatus.EM_ANALISE, "ALERTA_SEGURANCA_EM_ANALISE",
            RiskAlertStatus.FALSO_POSITIVO, "ALERTA_SEGURANCA_FALSO_POSITIVO",
            RiskAlertStatus.RESOLVIDO, "ALERTA_SEGURANCA_RESOLVIDO");

    private final RiskAlertRepository riskAlertRepository;
    private final ConversationRepository conversationRepository;
    private final ConversationMessageRepository messageRepository;
    private final CompanyRepository companyRepository;
    private final AuditService auditService;

    public RiskAlertService(RiskAlertRepository riskAlertRepository, ConversationRepository conversationRepository,
                             ConversationMessageRepository messageRepository, CompanyRepository companyRepository,
                             AuditService auditService) {
        this.riskAlertRepository = riskAlertRepository;
        this.conversationRepository = conversationRepository;
        this.messageRepository = messageRepository;
        this.companyRepository = companyRepository;
        this.auditService = auditService;
    }

    public record AlertaComConversa(RiskAlert alerta, Conversation conversation, String buyerCompanyName, String supplierCompanyName) {
    }

    @Transactional(readOnly = true)
    public List<AlertaComConversa> listarAlertas(RiskAlertStatus status) {
        List<RiskAlert> alertas = status != null
                ? riskAlertRepository.findByStatusOrderByCreatedAtDesc(status, PageRequest.of(0, 100))
                : riskAlertRepository.findByStatusInOrderByCreatedAtDesc(
                        List.of(RiskAlertStatus.ABERTO, RiskAlertStatus.EM_ANALISE), PageRequest.of(0, 100));
        if (alertas.isEmpty()) return List.of();

        Set<String> conversationIds = alertas.stream().map(RiskAlert::getConversationId).collect(java.util.stream.Collectors.toSet());
        Map<String, Conversation> convPorId = conversationRepository.findAllById(conversationIds).stream()
                .collect(java.util.stream.Collectors.toMap(Conversation::getId, c -> c));

        Set<String> companyIds = new java.util.HashSet<>();
        convPorId.values().forEach(c -> {
            companyIds.add(c.getBuyerCompanyId());
            companyIds.add(c.getSupplierCompanyId());
        });
        Map<String, String> nomePorId = companyRepository.findAllById(companyIds).stream()
                .collect(java.util.stream.Collectors.toMap(Company::getId, Company::getName));

        return alertas.stream().map(a -> {
            Conversation c = convPorId.get(a.getConversationId());
            return new AlertaComConversa(a, c, c == null ? null : nomePorId.get(c.getBuyerCompanyId()),
                    c == null ? null : nomePorId.get(c.getSupplierCompanyId()));
        }).toList();
    }

    public record ConversaSinalizada(Conversation conversation, List<ConversationMessage> messages, List<RiskAlert> alerts) {
    }

    /**
     * Acesso a UMA conversa sinalizada — exige um alerta (qualquer
     * estado, mesmo já resolvido: o histórico de uma conversa que já
     * teve alerta continua legítimo consultar). Sem alerta nenhuma vez,
     * nem o Suporte entra.
     */
    @Transactional
    public ConversaSinalizada acederConversaSinalizada(String conversationId, CurrentUser admin, HttpServletRequest req) {
        if (!riskAlertRepository.existsByConversationId(conversationId)) throw new NotFoundException("Conversa sinalizada");
        Conversation conversa = conversationRepository.findById(conversationId).orElseThrow(() -> new NotFoundException("Conversa sinalizada"));

        List<ConversationMessage> mensagens = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId);
        List<RiskAlert> alertas = riskAlertRepository.findByConversationIdOrderByCreatedAtDesc(conversationId);

        auditService.recordSafe(new AuditService.Entry(auditService.actorFrom(admin, req), "CONVERSA_SINALIZADA_ACEDIDA",
                "Conversation", conversationId, null, auditService.contextoFrom(req)));

        return new ConversaSinalizada(conversa, mensagens, alertas);
    }

    @Transactional
    public RiskAlert reclassificar(String alertId, CurrentUser admin, RiskAlertStatus status, String decision, HttpServletRequest req) {
        if (status == null || !STATUS_RECLASSIFICACAO.contains(status)) throw new ValidationException("Estado de alerta inválido.");
        RiskAlert alerta = riskAlertRepository.findById(alertId).orElseThrow(() -> new NotFoundException("Alerta de segurança"));

        alerta.setStatus(status);
        String decisaoCortada = decision == null ? null : decision.strip();
        if (decisaoCortada != null && decisaoCortada.length() > 1000) decisaoCortada = decisaoCortada.substring(0, 1000);
        alerta.setDecision(decisaoCortada == null || decisaoCortada.isBlank() ? null : decisaoCortada);
        alerta.setReviewedById(admin.id());
        alerta.setReviewedAt(Instant.now());

        Map<String, Object> detail = new LinkedHashMap<>(auditService.contextoFrom(req));
        detail.put("conversationId", alerta.getConversationId());
        detail.put("decision", alerta.getDecision());
        auditService.recordSafe(new AuditService.Entry(auditService.actorFrom(admin, req), ACTION_POR_STATUS.get(status),
                "RiskAlert", alertId, null, detail));

        return alerta;
    }
}
